KEP-4815+5234: DRA Update 4815 and split out 5234 for mixins

mortent opened the pull request:
<!-- short description of work done in PR e.g. updating milestone, adding new KEP, adding test requirements… -->  
- One-line PR description: Update ParitionableDevices KEP and create separate ResourceSliceMixins KEP

<!-- link to the k/enhancements issue -->
- Issue link: https://github.com/kubernetes/enhancements/issues/5234
- Issue link: https://github.com/kubernetes/enhancements/issues/4815

<!-- other comments or additional information -->
- Other comments: This updates 4815 to reflect the functionality that was actually implemented for 1.33. The mixins feature was originally part of 4815, but go cut from the scope for 1.33. So this moves that functionality into a separate KEP 5234.

mortent commented:
/wg device-management
/cc @pohly @johnbelamaric @klueska

bg-furiosa commented:
/lgtm

bg-chun commented:
/lgtm

pohly commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,853 @@
+- [x] (R) Production readiness review completed
+- [x] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
```
It is not clear to me when this checkbox is meant to be ticked off. Clearly not in the PR which creates a KEP as "implementable".

The wording above ("Check these off as they are completed for the Release Team to track.") implies that these boxes must be updated throughout the development cycle. I don't think that this is actually being done and enforced.

mortent replied:
Yeah, I'm not sure what is correct here. Looks like this is left unchecked in many KEPs.

kikisdeliveryservice replied:
You can leave this unchecked. We're reviewing whether this makes sense to have here.

pohly commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,853 @@
+
+With this KEP, DRA drivers can define metadata in mixins separately from specific
+devices and include them in a device by reference. This reduces the duplication
+in ResourceSlices and allows for more compact device definitions. 
```
Add:
> It also enables defining devices with more attributes and capacities than without mixins.
>
> Mixins can also be used in counter sets.

pohly commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,853 @@
+
+- Enable a more compact way to define devices in ResourceSlices so duplication can
+  be reduced and a larger number of devices can be published within a single
+  ResourceSlice.
```
Second goal:
- Enable defining devices with more attributes and capacities and counter sets with more counters.

Or is that just a side effect, not a goal?

pohly commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,853 @@
+
+1. The `DeviceCounterConsumption` field defines a list of named
+   `DeviceCounterConsumptionMixin`s. These define counters that can be
+   used to extend the counter consumption defined explicitly `DeviceCounterConsumption`
+   objects in the `ConsumesCounters` list on `Device`. The `CounterSet` from
```
```suggestion
   used to extend the counter consumption defined explicitly in `DeviceCounterConsumption`
   in the `ConsumesCounters` list on `Device`. The `CounterSet` from
```

pohly replied:
```suggestion
   used to extend the counter consumption defined explicitly in `DeviceCounterConsumption`
   in the `ConsumesCounters` list on `Device`. The `CounterSet` from
```

pohly commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,853 @@
+The mixins are referenced using the same pattern in all three places. The field
+is named `Includes` and will contain a list of references to the mixins. Rather
+than representing the references as a list of strings, there is a list of type
+`<SomeType>MixinRef` which has a `Name` field. This enables adding additional
```
So the YAML will be:
```
includes:
- name: something
```

Would it look better to avoid the generic "name" field name and instead use "mixin"?

```
includes:
- mixin: something
```

For someone who doesn't know what "includes" means it might provide an additional hint.

pohly replied:
The problem with just a `[]string` is that we have to be absolutely sure that we are never going to need anything other than a string for an entry. Using a struct is less risky.

Having said that, I agree that it seems safe here and simplifies the API.

mortent replied:
I've left it as it is for now. We can bet input from @johnbelamaric and the API reviewer here.

johnbelamaric replied:
I also do not have any ideas on why we would need a struct rather than a string. If we go with struct, I do prefer `mixin: foo` rather than `name: foo`

mortent replied:
I have updated the KEP use use strings instead of structs.

pohly commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,853 @@
+
+We have discussed adding a kubectl command or a plugin that will allow
+users to see the fully flattened versions of a ResourceSlice. But this
+is not in scope for alpha.
```
There is one more risk: by allowing devices and counter sets that have potentially more entries, the worst-case scenarios for scheduling becomes worse.

I think that's fine and we can add the usual disclaimer that "it depends on what DRA driver authors decide to use", which limits the potential for abuse because normal users can't trigger it.

mortent replied:
Updated

pohly commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,853 @@
+Mixins will no longer work when downgrading to a release without support for it.
+Since the mixins and include fields will be dropped, it means the scheduler will
+have an incorrect view of the ResourceSlice and might make incorrect allocation
+decisions.
```
This depends on how this feature gets implemented. If it is implemented the way I did it originally (= resolve mixin references when converting to the types in `k8s.io/dynamic-resource-allocation/api` which are then used by the scheduler), then we have two choices in that conversion code:
- drop devices with references when the feature is disabled
- convert devices even if disabled

In the first case, the scheduler will not make incorrect decisions because it never sees the devices. The downside is that there is no feedback to users - but we already have this problem overall ("why is the scheduler not picking this device or node here?").

In the second case, disabling the feature is incomplete: if some device causes abnormal scheduler performance, disabling the feature will not make that problem go away.

I prefer the first approach.

Either way, this needs to be discussed in the KEP under "implementation".

mortent replied:
I've added an "Implementation" section just after the description of the API and this is discussed there.

So I agree that the second approach is better. But it is not clear to me if we can support that? We can make the conversion logic aware of whether the feature is enabled in the scheduler (I think it is possible that this can be different from the APIServer?). But if the feature is disabled on the APIServer, the mixins fields will already have been dropped from the `v1beta1` types before it gets converted to the dynamic-resource-allocation types, so we won't know which devices to drop. Or am I missing something here?

pohly replied:
The feature gate in the scheduler disables the feature, regardless of whether the apiserver has devices with the additional feature.

> But if the feature is disabled on the APIServer, the mixins fields will already have been dropped from the v1beta1 types before it gets converted to the dynamic-resource-allocation types, so we won't know which devices to drop. Or am I missing something here?

If the device was created with the fields, the apiserver will send the fields even after the feature got turned off. That's the situation being discussed here because the scheduler then needs to do something.

The more problematic case is when a client sends them, the apiserver strips the fields, and then sends incomplete devices. There isn't much that we can do in this KEP about this - it's how Kubernetes APIs work.

pohly replied:
> So I agree that the second approach is better.

You meant "first approach", right? That is what is in "Implementation" now.

mortent replied:
Ah, so the APIServer only drops fields on writes and will return them regardless on reads. That does make sense and then my concern here doesn't apply.

Yeah, I meant the first approach.

pohly commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,853 @@
+
+During version skew where the apiserver supports the feature and the scheduler
+doesn't, the scheduler will see an incomplete view of the ResourceSlice. This
+means it might make incorrect allocation decisions.
```
Same here: it depends on the implementation.

mortent replied:
Updated, but the question in https://github.com/kubernetes/enhancements/pull/5238#discussion_r2056617587 applies here too.

pohly replied:
Commented there.

pohly commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,853 @@
+  - [Optional] Aggregation method:
+  - Components exposing the metric:
+- [ ] Other (treat as last resort)
+  - Details:
```
Thinking aloud: would it make sense to maintain some metrics which summarize which devices the scheduler sees?

Perhaps that can include a histogram of "attributes per device after expanding mixins" with "driver name" as one dimension.

A different metric could count mixin references. If that is zero, admins know that the feature is not in use and safe to disable.

mortent replied:
Are you thinking we should include this for alpha? We obviously have to dig into this in more detail before beta.

A metric with the number of devices sounds useful, in particular as it will be an important input to understand the scheduling performance. It probably have to be its own metric, since it wouldn't be a good dimension. A histogram with the attributes, capacity and counters also seems like interesting input, in particular as we have already mentioned that we think this could impact the worst-case scheduling performance.

pohly replied:
No, not for alpha. But it makes sense to start thinking about it.

mortent replied:
Yeah, I've started looking at updating [KEP-4816](https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/4816-dra-prioritized-list) for beta, so I'll be starting to think about metrics as part of that. But agree we need something here.

pohly commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,853 @@
+###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?
+
+Flattening the devices and counter sets will require slightly more work, but
+this is unlikely to have any meaningful impact on the time used for allocation.
```
... unless DRA driver authors start using really large counter sets.

mortent replied:
Yeah, I've updated the section to include this.

mortent commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -326,78 +321,42 @@ non-goals of this KEP.
-      will draw from the capacity pool when allocated. Therefore, if this capacity isn't
-      available in the pool, the device can not be allocated. Only references
-      to capacity pools in the same `ResourceSlice` is supported.
+   device can be allocated if the necessary counters (i.e. resources) is not
```
Updated.

mortent commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -2864,7 +1162,7 @@ End-to-end testing depends on a working resource driver and a container runtime
 driver](https://github.com/kubernetes/kubernetes/tree/master/test/e2e/dra/test-driver)
 was developed as part of the overall DRA development effort. We will extend
-this test driver to enable support for `PartitonableDevice`s and add tests to
+this test driver to enable support for `PartitionableDevice`s and add tests to
```
Yeah, I've tried to update the language. In some places it is not quite clear to me what "is correct".

mortent commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -2895,21 +1193,20 @@ drivers that they provide to customers.
-type. The new function will be offered through the newly added fields under `BasicDevice`.
-The kube-scheduler is expected to match the kube-apiserver minor version, 
+The API extensions proposed in this KEP are added as new fields on the
+`ResourceSliceSpec` and `Device` types. All the fields will be behind
```
I've updated the language in this section to avoid both present and future tense.

pohly commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,896 @@
+
+During version skew where the apiserver supports the feature and the scheduler
+doesn't, the devices that is using the mixins feature will be dropped and
+not visible to the scheduler (ref [Implementation](#implementation)).
```
This depends: with apiserver 1.34 + feature enabled and scheduler 1.33 (feature unknown), the apiserver sends devices with new fields, the scheduler doesn't know about the fields and processes incomplete devices.

This is only a problem for the initial alpha. For apiserver 1.35 + scheduler 1.34, the schedulers knows the fields and can do what is discussed under "Implementation".

So I think we are fine, we just need to document it a bit better.

mortent replied:
Yeah, that makes sense. I've updated the section to cover this scenario.

johnbelamaric commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/kep.yaml:
```diff
@@ -0,0 +1,51 @@
+  - "@thockin"
+approvers:
+  - "@mrunalp" # SIG-Node
+  - "@alculquicondor" # SIG-Scheduling
```
We need to find new approvers here. @dom4ha or @macsko perhaps?

dom4ha replied:
Yes, you can assign me.

johnbelamaric commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/kep.yaml:
```diff
@@ -0,0 +1,51 @@
+approvers:
+  - "@mrunalp" # SIG-Node
+  - "@alculquicondor" # SIG-Scheduling
+  - "@MaciekPytel" # SIG-Autoscaling
```
Same here. @jackfrancis or @raywainman maybe?

raywainman replied:
I would add @towca here who has been working the most on DRA stuff on the autoscaling side

jackfrancis replied:
I'll give it a review pass now (but yes either @towca or myself would make sense here as Maciek has transitioned away from active SIG Autoscaling duty)

mortent replied:
Added @jackfrancis as he has already reviewed it.

johnbelamaric commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,902 @@
+  be reduced and a larger number of devices can be published within a single
+  ResourceSlice.
+- Enable defining devices with more attributes, capacities, and consumed counters.
+- Enable defining counter sets with more counters.
```
Two things we should consider (not sure if these are goals or implementation choices): 1) enabling mixins to be per-pool not per-resource slice; 2) enabling counters to be per-pool not per-resource slice.

If necessary, these could be considered in beta. But I do think we're going to need them in time. The second may belong in partitionable not here.

mortent replied:
Yeah, we have an issue for that in https://github.com/kubernetes/kubernetes/issues/130785. We definitely need to make a decision on this in this cycle, as I think changing this must happen over two releases. I was hoping to handle this separately from this KEP, but open to including it here.

johnbelamaric commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,902 @@
+   These define counters that can be used to extend the counters
+   explicitly defined in a `CounterSet`. This allows for reduced duplication
+   if there are many identical physical devices that must be represented as
+   `CounterSet`s. `CounterSetMixins` can not be referenced directly by devices.
```
nit:
s/can not/cannot/

johnbelamaric commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,902 @@
+   only be referenced by devices.
+
+1. The `DeviceCounterConsumption` field defines a list of named
+   `DeviceCounterConsumptionMixin`s. These define counters that can be
```
I might tweak this a bit to have something like:

These define the pattern of consumption of counters, distinct from the specific underlying counter set from which they are being consumed.

Rather than "these define counters"

mortent replied:
Updated

johnbelamaric commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,902 @@
+If the APIServer is at 1.34 and the scheduler is at 1.33, the APIServer will send
+the new fields, but the scheduler will not know what to do about them. It will end
+up ignoring them, which can lead to incorrect scheduling decisions. Note that this
+scenario only applies to the initial 1.34 release and will not apply for 1.35+.
```
The recommendation is that the user should not enable this alpha feature unless the scheduler is updated to 1.34 and enables the alpha feature as well.

mortent replied:
Updated the language to include this.

mortent commented:
This KEP adds a few new limits on the size of slices/maps in the ResourceSlice in addition to the ones that were added as part of the Partitionable Devices and other features. But as I've tried a few scenarios I've realized the result is not great, as it makes it possible to add more attributes, capacity and counters through mixin without actually reducing the storage size of the ResourceSlice.

An example is that we currently limit the total number of counters across the counter sets in a ResourceSlice to 32. As a result, it is impossible to create a counter set with more than 32 counters. But with mixins, I can create a counter set mixin that is only referenced from a single counter set to create larger counter sets. This doesn't reduce the number of counters defined in a ResourceSlice, it just forces users to "abuse" mixins to bypass the limit.

We should set the limits based on the total number of attributes, capacity, and counters across the ResourceSlice, rather than based on whether they are defined in a Device or a Mixin. 

I suggest we set the limits to something like:
* Total combined number of attributes and capacity in a ResourceSlice is 4096 (32 * 128 devices)
* Total number of counters is 256
* Total number of consumed counters is 2048 (16 * 128 devices)

So no special limits for mixins, they count against the same limits as the properties defined in devices. With these limits the worst case size for the ResourceSlice increases from 1,107,864 bytes to 1,288,825 bytes as a result of adding mixins.

I think changing the limits for the counters should be pretty straightforward since those only affects fields that are still in alpha. So we can add the new ResourceSlice-wide limits and remove the more granular ones.
But I'm not sure if we will be able to easily remove the limit of 32 attributes+capacity per device, so I think this is something that would need to happen over two releases. But I think adding the ResourceSlice-wide limit of 4096 to cover both mixins and devices should be safe so we avoid any mixin-specific limits.

jackfrancis commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,902 @@
+  duplication between the published devices.
+* The size of the data required to specify each device reduces the number of
+  devices that can be defined in a single ResourceSlice without hitting the
+  Kubernetes size limit.
```
can we clarify which size limit we're up against (e.g., is it the payload of the `spec` data in a CRD?)

mortent replied:
Added language to clarify.

jackfrancis commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,902 @@
+mechanism for referencing mixins from devices and counter sets.
+
+A new `Mixins` field will be added to the `ResourceSliceSpec` as an
+optional field. It will have three properties, one for each of the three
```
should we say "...as an optional field of type `ResourceSliceMixins`

jackfrancis commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,902 @@
+#### More attributes, capacities and counters might worsen worst-case scheduling
+
+With mixins, DRA driver authors can choose to create more complex devices,
+which might lead to worse scheduling performance. But it is up to DRA driver
```
I might replace "But it is up to DRA driver authors to do this, so it is not something that can be triggered by normal users." with something like:

"This will not negatively effect existing scheduling performance of existing ResourceSlice definitions, but DRA driver authors taking advantage of mixins should be made aware of possible performance effects due to this increased referential complexity. Furthermore, this reinforces the criticality of ensuring that DRA primitives are optimized for maximum performance."

(That's my takeaway, at least.)

mortent replied:
I update the KEP to include your suggestion with some small edits.

jackfrancis commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,902 @@
+
+### API
+
+The exact set of proposed API changes can be seen below:
```
should we add a note that `...` after a type definition indicates that the proposed changes are additions to existing APIs?

jackfrancis commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,902 @@
+}
+
+// CounterSetMixin defines a mixin that a capacity pool can include.
+type CounterSetMixin struct {
```
nit: can we move this up so it immediately follows `ResourceSliceMixins` (where there is a reference to this new type)

jackfrancis replied:
I see that `ResourceSliceMixins` includes three new types and they are in that order, ignore!

jackfrancis commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,902 @@
+// The main purposes of these mixins is to reduce the memory footprint
+// of devices since they can reference the mixins provided here rather
+// than duplicate them.
+type ResourceSliceMixins struct {
```
Sorry if this initiates a lengthy, annoying naming debate! But: because this is a struct would it be preferable to call this `ResourceSliceMixinSpec` instead (and then we'd ostensibly update the changes to `ResourceSliceSpec` to include a new `MixinSpec` property (instead of a `Mixins` property).

mortent replied:
Interesting question, but when I look at the existing fields on the `ResourceSliceSpec`, they don't use the `...Spec` naming convention in the field name or the type. So for consistency, I think we should just follow the pattern and keep the current names.

jackfrancis commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,902 @@
+### Implementation
+
+The DRA scheduler plugin will flatten the counter sets and devices before
+going through the allocation process. This will happen as part of conversion
```
Putting my SIG Autoscaling hat on, is this a sufficient description of the plan to provide a standard interface for rendering the "flattened" ResourceSlice (after following and processing all mixin references)? In the worst case scenario those conversions happen surgically throughout various parts of the k/k codebase, which would make it hard for downstream components like cluster-autoscaler and karpenter to plumb into durable, reusable libraries.

mortent replied:
Yeah, I think this is a good question. Every tool that needs to understand the full device definitions will need to flatten the mixins, but the suggested implementation here doesn't lend itself easily to reuse as the flattening happens as part of conversion into the scheduler-specific format. 
I added a separate section under "Risks and Mitigations" for this. The most obvious solution here is that we provide a library that handles the flattening, although I'm not sure which types (most likely v1beta2) we should do this for.

jackfrancis replied:
cc @towca

mortent replied:
I'm almost certain that with the current integration the Cluster Autoscaler is already using the conversion logic from the `v1beta1` API to the internal dynamic-resource-allocation version used by the DRA scheduler plugin. So it shouldn't be a concern for that integration.

But for other uses, like implementing a view of the fully flattened devices in kubectl, it is less clear how flattening as part of conversion can be reused.

towca replied:
> I'm almost certain that with the current integration the Cluster Autoscaler is already using the conversion logic from the `v1beta1` API to the internal dynamic-resource-allocation version used by the DRA scheduler plugin. So it shouldn't be a concern for that integration.

I don't think that's true, CA integrates with scheduler through [`SharedDRAManager`](https://github.com/kubernetes/kubernetes/blob/074b21e76ca4974b6cf9574022ae4ea7c609f675/pkg/scheduler/framework/listers.go#L112C6-L112C22), which operates on `v1beta1` objects. From a quick look through the DRA plugin code, it doesn't seem to convert the objects further? Could you point me to the internal version?

FWIW this shouldn't impact most of the integration as CA doesn't have to flatten the slices for most purposes. But it does have to do it e.g. for calculating utilization for scale-down. Having a library to use for it would be preferrable. I'd also call out the work required in CA to integrate with this KEP somewhere.

jackfrancis commented:
> This KEP adds a few new limits on the size of slices/maps in the ResourceSlice in addition to the ones that were added as part of the Partitionable Devices and other features. But as I've tried a few scenarios I've realized the result is not great, as it makes it possible to add more attributes, capacity and counters through mixin without actually reducing the storage size of the ResourceSlice.
> 
> An example is that we currently limit the total number of counters across the counter sets in a ResourceSlice to 32. As a result, it is impossible to create a counter set with more than 32 counters. But with mixins, I can create a counter set mixin that is only referenced from a single counter set to create larger counter sets. This doesn't reduce the number of counters defined in a ResourceSlice, it just forces users to "abuse" mixins to bypass the limit.
> 
> We should set the limits based on the total number of attributes, capacity, and counters across the ResourceSlice, rather than based on whether they are defined in a Device or a Mixin.
> 
> I suggest we set the limits to something like:
> 
> * Total combined number of attributes and capacity in a ResourceSlice is 4096 (32 * 128 devices)
> * Total number of counters is 256
> * Total number of consumed counters is 2048 (16 * 128 devices)
> 
> So no special limits for mixins, they count against the same limits as the properties defined in devices. With these limits the worst case size for the ResourceSlice increases from 1,107,864 bytes to 1,288,825 bytes as a result of adding mixins.
> 
> I think changing the limits for the counters should be pretty straightforward since those only affects fields that are still in alpha. So we can add the new ResourceSlice-wide limits and remove the more granular ones. But I'm not sure if we will be able to easily remove the limit of 32 attributes+capacity per device, so I think this is something that would need to happen over two releases. But I think adding the ResourceSlice-wide limit of 4096 to cover both mixins and devices should be safe so we avoid any mixin-specific limits.

Great initial thoughts, I would go ahead and move your thinking and preliminary conclusions into the KEP where it will probably get the most engagement.

pohly commented:
> But I'm not sure if we will be able to easily remove the limit of 32 attributes+capacity per device, so I think this is something that would need to happen over two releases.

Yes, this would need ratcheting.

We have already gradually moved away from individual per-slice and per-map limits towards aggregating at higher levels. You proposal is now basically to move this up to the root level of the entire slice. This makes sense to me and @thockin has approved the previous aggregated API limits, but it still is a bit unusual. Therefore I would like to hear from others what they think about taking this approach to the logical conclusion.

dom4ha commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,902 @@
+
+### Implementation
+
+The DRA scheduler plugin will flatten the counter sets and devices before
```
By flattening we still risk that large number of references combined with a large size of mixing could cause scheduler OOM, as there would be no mechanism of keeping the consumption under control.

I think the in-memory representation should stay as is, but the allocator should iterate over mixins somehow. Is that feasible?

mortent replied:
We could do that, but it means additional work to dereference the mixins every time they are needed and more complexity in the allocator to handle it.

I also think the question around memory usage in the scheduler goes beyond just mixins. The memory usage per device does matter, but so does the number of devices. Currently we allow a maximum of 127 devices per ResourceSlice, but there is no other limit on the number of ResourceSlices than the number of objects for a single type in Kubernetes. We can make changes to the allocator to make sure we don't try to keep all devices in memory at the same time, but that comes with other challenges.

We probably should look at whether we should place a limit on the number of devices in a cluster and then see what kind of impact that has on the memory usage of the scheduler.

mortent replied:
After taking a look at the code, I do think handling ResourceSlices unflattened in the allocator should be very doable. For counters in either counter sets or device counter consumption, we just need to walk the mixins and we can probably turn those into an appropriate data structure for that as part of conversion. For device mixins, we do need to flatten it at some point in order to evaluate the CEL expressions for the selectors. But we already do an [extra conversion](https://github.com/kubernetes/kubernetes/blob/5290c24e850baf50b6871601475d0268ecff3fba/staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go#L931-L940) from the internal dynamic resource allocation API to the `v1beta1` (eventually `v1beta2`) API for CEL evaluation, and we also [prep the attributes and capacity maps for CEL](https://github.com/kubernetes/kubernetes/blob/5290c24e850baf50b6871601475d0268ecff3fba/staging/src/k8s.io/dynamic-resource-allocation/cel/compile.go#L231-L251). I think both places can be used to flatten the mixins for CEL evaluation.

So I think we have alternatives if flattening devices in the `v1beta1` to dynamic-resource-allocation conversion causes the per-device memory usage to be too high.

As for the total number of devices, it seems like these are managed through an informer in the DRAManager. But I see that we only do conversion for devices that are available for the specific node being processed in the allocator. The question of whether there are enough safeguards to make sure a large number of ResourceSlices doesn't cause the scheduler to OOM might be a question for the GA process of https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/4381-dra-structured-parameters.

@pohly

pohly replied:
> we just need to walk the mixins and we can probably turn those into an appropriate data structure for that as part of conversion

Or even better, implement that walking as part of the CEL variable lookup. There's an [open TODO](https://github.com/kubernetes/kubernetes/blob/c519248e8a865d837f3f40308eaf9559e605306d/staging/src/k8s.io/dynamic-resource-allocation/cel/compile.go#L229-L230). I expect performance to get better because it avoids memory allocations.

Speaking of CEL... there is one aspect of mixins that we hadn't considered. No matter how we specify the limits for the resource slice itself, we now can have more than the traditional 32 attributes/quantities per device. On top of that, looking them up becomes more costly. I think we can ignore the extra lookup cost (cost estimates are just that - estimates), but the higher number of attributes has a real impact on validation because the worst case size of the maps becomes higher, which makes the estimated cost of **the same** expression higher than it was before. If an expression was just below the cost limit in 1.32, then it might be above the cost limit in 1.33, which breaks users.

We could increase the cost limit in 1.33, but I believe it is impossible to predict by how much. A complex expression might have O(32 * 32) squared costs now and O(64 * 64) if we double the number of allowed attributes/quantities, so doubling the cost limit would not be enough.

I don't see a good solution. Perhaps we should make increasing the number of allowed attributes/quantities a non-goal? We only added it because I asked, not because that really was the intent. So perhaps it is not a big loss if we don't do it? If we go down that route, then in addition to whatever aggregate limits we have for a valid slice we also have to validate that after resolving mixings, the number of attributes/quantities is not higher than what was allowed before.

mortent replied:
Yeah, resolving the mixins and validate the number of attributes and capacities during validation seems reasonable. We are already doing most of the work to accommodate this anyway. I like that approach. If there are actual requirements to support a higher number of attributes/capacities for a device, that is a different discussion than mixins.

Only caveat is that if we want to implement https://github.com/kubernetes/kubernetes/issues/130785 as @johnbelamaric brought up in https://github.com/kubernetes/enhancements/pull/5238#discussion_r2058889372, we won't be able to validate until allocation time. But quite a bit of validation will have to happen at allocation time if we do this.

mortent replied:
@pohly I think the one remaining question from this discussion that haven't been addressed yet is whether a large number of ResourceSlices in a cluster could cause the scheduler to OOM. Is there already a safe-guard for this in the informers used in the DRAManager? And do we need any additional protection to make sure the allocator doesn't convert so many ResourceSlices that it becomes an issue?

dom4ha replied:
It's great we don't have to flatten everything.

dom4ha commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,902 @@
+the devices that they manage in ResourceSlices. This information is used by the
+scheduler when selecting devices for user requests in ResourceClaims.
+
+With this KEP, DRA drivers can define metadata in mixins separately from specific
```
I suspect this topic might have been discussed already, but why don't we use a word templates instead of mixins? Wouldn't it be more straightforward naming convention that would be easier to understand?

mortent replied:
Not sure what you mean by word templates here. Could you provide an example?

dom4ha replied:
I meant:
`DeviceMixin` -> `DeviceTemplate`
but after giving it a second thought,  use of 'mixin' seems more appropriate.

dom4ha commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,902 @@
+  the full definition for a specific device, so this might be added to the scope
+  for Beta or GA.
+
+## Proposal
```
Can you describe how conflicts in flattening would be resolved? Is overwriting allowed and if not, how errors would be reported?

mortent replied:
This is described in the comments on the `Includes` fields on `CounterSet`, `Device`, and `DeviceCounterConsumption` types. I think that is the right place to document this, as the order of the mixins listed matters for conflicting properties.

mortent replied:
Added.

dom4ha commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,902 @@
+type ResourceSliceSpec struct {
+  ...
+
+  // Mixins defines the mixins available for devices and counter sets
```
Can you expand the comment to clearly define the purpose of mixins and how they will be merged with other attributes (how possible conflicts are handled)

mortent replied:
This is described in the comments on the `Includes` fields on `CounterSet`, `Device`, and `DeviceCounterConsumption` types. I think that is the right place to document this, as the order of the mixins listed matters for conflicting properties.

dom4ha reviewed (commented):
Thanks for splitting these KEPs, as it's more self-contained and it's much easier to understand this enhancement now.

mortent commented:
> Great initial thoughts, I would go ahead and move your thinking and preliminary conclusions into the KEP where it will probably get the most engagement.

Updated the KEP with the proposal for limits with a new section under the Design Details. I've left the Partitionable Devices KEP as is so it reflects what was implemented for 1.33. If these changes get accepted and implemented, I will also update the Partitionable Devices KEP.

mortent commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,853 @@
+  // The mixins referenced here must be defined in the same
+  // ResourceSlice.
+  //
+  // The maximum number of includes is 8.
```
As discussed in comments further down, the KEP has been updated to propose larger changes to how we set the limits.

mortent commented:
The KEP has been updated based on the comments.

dom4ha commented:
/approve
It looks good from the scheduler perspective

johnbelamaric reviewed (commented):
/approve

Some non-blocking nits

pohly commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -647,441 +475,175 @@ type Device struct {
-  // CapacityPool defines the capacity pool from which the
-  // capacities defined (either directly or through a mixin)
-  // will be consumed from.
+// Counter describes a quantity associated with a device.
```
But not just one particular device as in "device defined in the ResourceSlice", right?

```suggestion
// Counter describes a quantity that multiple devices consume when in use.
```

It was tempting to say "quantity of something" but that just raises more questions than it answers, so I left out "of something".

pohly commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -3117,12 +1420,29 @@ No.
+  is limited to 1024.
+
+The `ResourceSlice`-wide limits on fields within the `Device` object is used to allow users
+to decide whether to have few devices with many properties or many devices with few properties.
```
How much does the worst-case ResourceSlice size grow now with these limits?

We traditionally haven't gone into those details in KEPs, but I think we can answer this already and it would be good to document.

mortent replied:
I've added the data about the increase in the worst-case ResourceSlice size.

pohly commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,912 @@
+devices and include them in a device by reference. This reduces the duplication
+in ResourceSlices and allows for more compact device definitions. It also enables
+defining devices with more attributes, capacities and consumed counters than without
+mixins. Mixins can also be used in counter sets.
```
We dropped allowing "more attributes, capacities", didn't we?

mortent replied:
Yeah, I updated the goals to reflect that, but missed this one. Updated.

pohly commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,912 @@
+  users see the flattened device definitions. Mixins does make it harder to find
+  the full definition for a specific device, so this might be added to the scope
+  for Beta or GA.
+- Enable devices to have more than 32 attributes and capacities. Increasing this
```
```suggestion
- Enable devices to have more than 32 attributes and capacities. Increasing this would
```

pohly commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,912 @@
+ResourceSlice definitions, but DRA driver authors taking advantage of mixins should
+be made aware of possible performance effects due to this increased referential complexity.
+
+This also demonstrates that DRA driver authors should consider performancer when they
```
```suggestion
This also demonstrates that DRA driver authors should consider performance when they
```

pohly commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,912 @@
+  //
+  // The maximum number of attributes and capacities across all devices
+  // and device mixins in a ResourceSlice is 4096. When flattened, the
+  // total number of attributes and capacities for each device can not
```
```suggestion
  // total number of attributes and capacities for each device must not
```

pohly commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,912 @@
+  //
+  // The maximum number of attributes and capacities across all devices
+  // and device mixins in a ResourceSlice is 4096. When flattened, the
+  // total number of attributes and capacities for each device can not
```
```suggestion
  // total number of attributes and capacities for each device must not
```

pohly commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,912 @@
+* Total number of counters is 256.
+* Total number of consumed counters is 2048 (so with the maximum number of devices, there can be 16 per device).
+
+We will still enforce some per-slice limits:
```
```suggestion
We will still enforce some per-field limits:
```

My initial read expanded "per-slice" to "per-ResourceSlice", which was the opposite of what was meant...

mortent replied:
Yeah, agree that "per-slice" is a bit confusing. Updated.

pohly commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,912 @@
+
+The limits on the number of counters across counter sets, mixins and device counter consumption in 1.33 for the
+Partitionable Devices KEP will be removed, as those are still in alpha.
+The limit of 32 on the number of attributes and capacities per device will be removed over the next 2 releases (1.34 and 1.35) to
```
I'm not sure about this. The problem I mentioned in https://github.com/kubernetes/enhancements/pull/5238#discussion_r2062707393 ("makes the estimated cost of the same expression higher than it was before" and therefore invalid) remains, even if we do this gradually. We would have to increase the cost limit, but don't know by how much.

mortent replied:
Yeah, I've updated the language here. I think we essentially just want to change the validation to do the check with mixins applied rather than on just the device. That should be a safe change I think, since all devices that fail the current check will also fail the updated one.

pohly commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,912 @@
+We expect no non-infra related flakes in the last month as a GA graduation criteria.
+-->
+
+E2e tests will be added to verify that the mixins are properly flattened and
```
```suggestion
E2e tests will be added to verify that the mixins are properly applied and
```

pohly commented on keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md:
```diff
@@ -0,0 +1,912 @@
+## Implementation History
+
+- 1.33: first KEP revision as part of the Partitionable Devices KEP
+- 1.34: split out into a separate KEP and initial implementation.
```
This 1.34 entry isn't "history" yet. If we merge this now and then don't follow through in 1.34, we'll have an incorrect statement in the KEP. Perhaps add `(tentative)`?

johnbelamaric replied:
If we're being picky, the splitting is history (with this merge). But the initial implementation is not, so maybe we just leave that bit off.

mortent replied:
Yeah, I have removed the implementation bit.

pohly reviewed (approved):
/lgtm
/hold

In case that we want additional (API?) reviews.

sanposhiho commented:
/cc

johnbelamaric commented:
/hold cancel
